### PR TITLE
Matrix Solve Op enabled for ROCm

### PR DESCRIPTION
**Pull request by stevenireeves**
Added necessary code and functions for a ROCm based MatrixSolveOp.

### CODE REVIEW

File: tensorflow/core/util/rocm_solvers.h [hunk 1]
```diff
@@ -44,14 +44,14 @@ struct ROCmComplexT {
...
+  typedef rocblas_float_complex type;
 };
 template <>
 struct ROCmComplexT<std::complex<double>> {
-  typedef hipDoubleComplex type;
+  typedef rocblas_double_complex type;
 };
 // Converts pointers of std::complex<> to pointers of
-// cuComplex/cuDoubleComplex. No type conversion for non-complex types.
+// ROCmComplex/ROCmDoubleComplex. No type conversion for non-complex types.
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.h]: nit: did you mean to say `rocblas_float_complex / rocblas_double_complex` ?

File: tensorflow/core/kernels/linalg/matrix_solve_op.cc [hunk 2]
```diff
@@ -364,13 +369,263 @@ class MatrixSolveOpGpu : public AsyncOpKernel {
   bool adjoint_;
 };
 
+#elif TENSORFLOW_USE_ROCM
+template <class Scalar>
+class MatrixSolveOpGpu : public AsyncOpKernel {
```
Comment (deven-amd) [inline tensorflow/core/kernels/linalg/matrix_solve_op.cc]: The code for `class MatrixSolveOpGpu` has a lot of duplication from the `#if GOOGLE_CUDA` side. In order to be able to upstream this, we will need to combine them into one, and `#if GOOGLE_CUDA ... #elif TENSORFLOW_USE_ROCM ... #endif ` in places where the code differs.

File: tensorflow/core/util/rocm_solvers.h [hunk 3]
```diff
@@ -78,8 +88,62 @@ class ROCmSolver {
...
+  // Wrappers for ROCSolver start here
+  //
+  // The method names below
+  // map to those in ROCSolver, which follow the naming
+  // convention in LAPACK see
+
+  // LU factorization.
+  // Computes LU factorization with partial pivoting P * A = L * U.
+  template <typename Scalar>
+  Status getrf(int m, int n, Scalar* dev_A, int lda, int* dev_pivots, int* info);
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.h]: Can we try to keep the API name + interface, common between `ROCmSolver` / `CudaSolver` classes. 

Ideally we would want to have only one class + header file ( `GpuSolver` +  `gpu_solver.h` ) and platform specific   implementations for them ( in `rocm_solver.cpp` and `cuda_solver.cpp` ). There is a good chance we will be requested to do this too when we try to upstream this change.

File: tensorflow/core/util/rocm_solvers.h [hunk 4]
```diff
@@ -78,8 +88,62 @@ class ROCmSolver {
...
+
+  // Uses LU factorization to solve A * X = B.
+  template <typename Scalar>
+  Status getrs(const rocblas_operation trans, int n, int nrhs, Scalar* A,
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.h]: same as above

File: tensorflow/core/util/rocm_solvers.h [hunk 5]
```diff
@@ -78,8 +88,62 @@ class ROCmSolver {
...
+               int lda, const int* dev_pivots, Scalar* B, int ldb);
+
+  template <typename Scalar>
+  Status
+  getrf_batched(int m, int n, Scalar** dev_A, int lda, int* dev_pivots,
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.h]: same as above

File: tensorflow/core/util/rocm_solvers.h [hunk 6]
```diff
@@ -78,8 +88,62 @@ class ROCmSolver {
...
+                rocblas_stride stride, DeviceLapackInfo* info, const int batch_count);
+
+  template <typename Scalar>
+  Status getrs_batched(const rocblas_operation trans, int n, int nrhs,
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.h]: same as above

File: tensorflow/core/util/rocm_solvers.h [hunk 7]
```diff
@@ -154,6 +218,59 @@ class ScratchSpace {
...
+  return GetScratchSpace<Scalar>(TensorShape({size}), debug_info, on_host);
+}
+
+inline DeviceLapackInfo ROCmSolver::GetDeviceLapackInfo(
+    int64 size, const std::string& debug_info) {
+  DeviceLapackInfo new_dev_info(context_, size, debug_info);
+  scratch_tensor_refs_.emplace_back(new_dev_info.tensor());
+  return new_dev_info;
+}
+
```
Comment (deven-amd) [inline tensorflow/core/util/rocm_solvers.h]: this looks to be the same as the contents in `cuda_solver.h`.... we can eliminate this duplication too, if we unify the two headers into one

**Comment by stevenireeves** [comment]
@deven-amd can you look again?

File: tensorflow/core/kernels/unique_op_gpu.cu.h [hunk 8]
```diff
@@ -29,13 +29,12 @@ limitations under the License.
 #include "tensorflow/core/lib/core/bits.h"
 #include "tensorflow/core/lib/core/status.h"
 #include "tensorflow/core/util/gpu_kernel_helper.h"
+#include "tensorflow/core/util/gpu_solvers.h" // Fore ScratchSpace
```
Comment (stevenireeves) [inline tensorflow/core/kernels/unique_op_gpu.cu.h]: Fix typo

File: tensorflow/core/util/cuda_solvers.cc [hunk 9]
```diff
@@ -965,7 +965,7 @@ static inline Status TrsvImpl(SolverFnT solver, cublasHandle_t cublas_handle,
 
 #define TRSV_INSTANCE(Scalar, type_prefix)                                   \
   template <>                                                                \
-  Status CudaSolver::Trsv<Scalar>(                                           \
+  Status GpuSolver::Trsv<Scalar>(                                           \
```
Comment (deven-amd) [inline tensorflow/core/util/cuda_solvers.cc:968]: make sure you run `clang-format` on your changes, when you file the upstream PR

File: tensorflow/core/util/gpu_solvers.h [hunk 10]
```diff
@@ -190,6 +220,43 @@ class CudaSolver {
...
+  Status GetrsBatched(const rocblas_operation trans, int n, int nrhs,
+                      Scalar** A, int lda, int* dev_pivots,
+                      rocblas_stride stride, Scalar** B, const int ldb,
+                      const int batch_count);
+
+
+  template <typename Scalar>
+  Status Trsm(rocblas_side side, rocblas_fill uplo, rocblas_operation trans,
+              rocblas_diagonal diag, int m, int n, const Scalar* alpha,
+              const Scalar* A, int lda, Scalar* B, int ldb);
```
Comment (deven-amd) [inline tensorflow/core/util/gpu_solvers.h:258]: when you file the upstream PR, see if you can come up with a way to keep the interface/signature for these functions the same for both CUDA and ROCm. My guess is that the upstream reviewers will request for this

**Review by deven-amd — APPROVED** [review]
When you file the upstream PR for this, have the changes related to the file-renaming in a separate commit. That way, if they reject that (file-renaming) part, it is easy to drop it. Also makes it easier to review

Also please double check the unit-test part...I would have expected to see either some `no_rocm` tags getting removed or the removal `if built_with_rocm() : skip` references within unit tests. This PR does not include any change on the unit-test side!

**Comment by stevenireeves** [comment]
> Also please double check the unit-test part...I would have expected to see either some `no_rocm` tags getting removed or the removal `if built_with_rocm() : skip` references within unit tests. This PR does not include any change on the unit-test side!

If you take a look at the original implementation for matrix_solve_op, it was #if GOOGLE_CUDA, or else CPU. I believe this is the reason why GPU tests were enabled, when I made the rocm side. I'll double check to make sure nothing is missing.

**Pull request merged**